### PR TITLE
feat(tracer): specify subsegment name when capturing class method

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

Community member @misterjoshua suggested a feature in #1084 to allow customers to set a custom name for the subsegment generated by the `Tracer.captureMethod` method.

At the moment Tracer automatically sets the subsegment name to the same value of the name of the method. This is a sensible choice but in some cases, when using non descriptive method names, customers might want to set their own names.

This PR adds an optional parameter to the options for `captureMethod` so that the custom name can be set. The PR also splits the type used for the options in `captureMethod` and `captureLambdaHandler` so that the new parameter is valid only for the former (the latter will always be called `## index.${HANDLER NAME}` for now).

The PR also adds unit test cases as well as e2e test cases and does some light housekeeping around the e2e tests and docstrings related to this feature.

### How to verify this change

Check out newly added unit test cases and new e2e test cases (results [here](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/3082996067)).

### Related issues, RFCs

**Issue number:** #1084

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

N/A

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/core/tracer.md [hunk 1]
```diff
@@ -256,10 +256,11 @@ You can trace other Class methods using the `captureMethod` decorator or any arb
     }
      
     const handlerClass = new Lambda();
-    export const handler = handlerClass.handler.bind(handlerClass); // (1)
+    export const handler = handlerClass.handler.bind(handlerClass); // (2)
     ```
 
-    1. Binding your handler method allows your handler to access `this`.
+    1. You can set a custom name ofr the subsegment by passing `subsegmentName` to the decorator, like: `@tracer.captureMethod({ subSegmentName: '### myCustomMethod' })`.
```
Comment (misterjoshua) [inline docs/core/tracer.md:262]:
```suggestion
    1. You can set a custom name for the subsegment by passing `subSegmentName` to the decorator, like: `@tracer.captureMethod({ subSegmentName: '### myCustomMethod' })`.
```

File: packages/tracer/src/types/Tracer.ts [hunk 2]
```diff
@@ -29,21 +29,70 @@ type TracerOptions = {
...
+ * ```typescript
+ * const tracer = new Tracer();
+ *
+ * class Lambda implements LambdaInterface {
+ *   @tracer.captureMethod({ subSegmentName: 'gettingChargeId', captureResponse: false })
+ *   private getChargeId(): string {
+ *     return 'foo bar';
+ *   }
+ *
+ *   @tracer.captureLambdaHandler({ subSegmentName: '', captureResponse: false })
```
Comment (misterjoshua) [inline packages/tracer/src/types/Tracer.ts:84]: Should `subSegmentName` be in this `@tracer.captureLambdaHandler()` example?

**Review by misterjoshua — COMMENTED** [review]
Just a few small nitpicks in the docs. Otherwise, this looks good to me.

File: packages/tracer/src/types/Tracer.ts (hunk 2 above)
Comment (dreamorosi) [reply packages/tracer/src/types/Tracer.ts:84]:
```suggestion
 *   @tracer.captureLambdaHandler({ captureResponse: false })
```

**Comment by dreamorosi** [comment]
> Just a few small nitpicks in the docs. Otherwise, this looks good to me.

Thank you for taking a look at this and good spotting the mistakes in the docstrings, I have committed your suggestions :)

**Review by ijemmy — DISMISSED** [review]

File: packages/tracer/tests/e2e/allFeatures.decorator.test.functionCode.ts [hunk 3]
```diff
@@ -9,8 +9,8 @@ const serviceName = process.env.EXPECTED_SERVICE_NAME ?? 'MyFunctionWithStandard
 const customAnnotationKey = process.env.EXPECTED_CUSTOM_ANNOTATION_KEY ?? 'myAnnotation';
 const customAnnotationValue = process.env.EXPECTED_CUSTOM_ANNOTATION_VALUE ?? 'myValue';
 const customMetadataKey = process.env.EXPECTED_CUSTOM_METADATA_KEY ?? 'myMetadata';
-const customMetadataValue = JSON.parse(process.env.EXPECTED_CUSTOM_METADATA_VALUE) ?? { bar: 'baz' };
-const customResponseValue = JSON.parse(process.env.EXPECTED_CUSTOM_RESPONSE_VALUE) ?? { foo: 'bar' };
+const customMetadataValue = process.env.EXPECTED_CUSTOM_METADATA_VALUE ? JSON.parse(process.env.EXPECTED_CUSTOM_METADATA_VALUE) : { bar: 'baz' };
+const customResponseValue = process.env.EXPECTED_CUSTOM_RESPONSE_VALUE ? JSON.parse(process.env.EXPECTED_CUSTOM_RESPONSE_VALUE) : { foo: 'bar' };
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/tracer/tests/e2e/allFeatures.decorator.test.functionCode.ts:13]: Parsing a non-JSON / `undefined` value throws an error rather than evaluate to false, so if the goal was to use the default value (after the `??`) when one is not explicitly provided in the env, then this should be correct.

image

Happy to revert if you think it's not ok or I misunderstood the intent of the original expression.

[the remark this answers is not in the document]
Comment (ijemmy) [reply packages/tracer/tests/e2e/allFeatures.decorator.test.functionCode.ts:13]: No. I just want to check that it's really a missing edge case.

**Review by ijemmy — APPROVED** [review]

**Review by flochaz — APPROVED** [review]
